File: packages/java/src/main/java/com/digitalt3/commons/api/Timer.java
package com.digitalt3.commons.api;

/**
 * Single-use monotonic timer associated with a {@link Logger}.
 *
 * <p>Start a timer with {@link #start()}, then complete it once using
 * {@link #stop()} or {@link #finish()}. Completion emits an INFO canonical
 * event with the configured event name and a non-negative {@code duration.ms}
 * through the owning logger's normal event pipeline.</p>
 *
 * @since 0.1.0
 */
public interface Timer {

    // PUBLIC_INTERFACE
    /**
     * Start this timer.
     *
     * @return this timer
     * @throws IllegalStateException if the timer has already started or the logger is closed
     */
    Timer start();

    // PUBLIC_INTERFACE
    /**
     * Stop this timer, record its duration, and emit its completion event.
     *
     * @return the measured non-negative duration in milliseconds
     * @throws IllegalStateException if the timer has not started, has already completed,
     *     or the logger is closed
     */
    long stop();

    // PUBLIC_INTERFACE
    /**
     * Complete this timer as an alias for {@link #stop()}.
     *
     * @return the measured non-negative duration in milliseconds
     * @throws IllegalStateException if the timer has not started, has already completed,
     *     or the logger is closed
     */
    long finish();

    // PUBLIC_INTERFACE
    /**
     * Return the measured duration after this timer has completed.
     *
     * @return the measured non-negative duration in milliseconds
     * @throws IllegalStateException if this timer has not completed
     */
    long durationMs();
}

File: packages/node/src/sdk/Timer.ts
import { Timer, TimerContext } from '../api/Timer';
import { Logger } from '../api/Logger';

/**
 * Internal logger capability used to preserve normal pipeline and lifecycle behavior.
 */
interface TimerLogger extends Logger {
  info(message: string, context?: Record<string, unknown>): void;
  ensureTimerLoggerOpen(): void;
}

/**
 * Measure elapsed monotonic time and emit one canonical completion event.
 */
export class TimerImpl implements Timer {
  private startedAt?: bigint;
  private stopped = false;

  /**
   * Create an unstarted timer.
   *
   * @param logger - The logger that emits the completion event.
   * @param name - Canonical completion event name.
   * @param context - Optional completion-event metadata.
   */
  constructor(
    private readonly logger: TimerLogger,
    private readonly name: string,
    private readonly context: TimerContext = {},
  ) {
    if (typeof name !== 'string') {
      throw new TypeError('name must be a string');
    }
    if (name.trim().length === 0) {
      throw new Error('name must not be blank');
    }
  }

  // PUBLIC_INTERFACE
  /**
   * Start this timer.
   *
   * @returns This timer for fluent usage.
   * @throws Error if already started or the associated logger is closed.
   */
  public start(): TimerImpl {
    if (this.startedAt !== undefined) {
      throw new Error('Timer has already been started');
    }

    this.logger.ensureTimerLoggerOpen();
    this.startedAt = process.hrtime.bigint();
    return this;
  }

  // PUBLIC_INTERFACE
  /**
   * Stop this timer, emit its completion event, and return elapsed milliseconds.
   *
   * @returns The non-negative elapsed duration in milliseconds.
   * @throws Error if not started, already stopped, or the logger is closed.
   */
  public stop(): number {
    if (this.startedAt === undefined) {
      throw new Error('Timer has not been started');
    }
    if (this.stopped) {
      throw new Error('Timer has already been stopped');
    }

    this.logger.ensureTimerLoggerOpen();
    const elapsedMs = Number(process.hrtime.bigint() - this.startedAt) / 1_000_000;
    const completionContext: Record<string, unknown> = {
      ...this.context,
      'event.name': this.name,
      'duration.ms': elapsedMs,
    };

    this.logger.info(`${this.name} completed`, completionContext);
    this.stopped = true;
    return elapsedMs;
  }

  // PUBLIC_INTERFACE
  /**
   * Finish this timer.
   *
   * @returns The non-negative elapsed duration in milliseconds.
   */
  public finish(): number {
    return this.stop();
  }
}
